<a href="https://colab.research.google.com/github/AndreaRR-2003/03MIAR----Algoritmos-de-Optimizacion---2026/blob/main/SEMINARIO%5CSeminario_Algoritmos.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Algoritmos de optimización - Seminario<br>
nombre y Apellidos: Anna Muñoz Bono y Andrea Rojo Rodríguez  <br>
Url: https://github.com/.../03MAIR---Algoritmos-de-Optimizacion---2019/tree/master/SEMInARIO<br>
Problema:
> 1. Sesiones de doblaje <br>
>2. Organizar los horarios de partidos de La Liga<br>
>3. Combinar cifras y operaciones

Descripción del problema:

Problema 2. Organizar los horarios de partidos de La Liga(I)

• Desde la La Liga de fútbol profesional se pretende organizar los horarios de los partidos de
liga de cada jornada. Se conocen algunos datos que nos deben llevar a diseñar un
algoritmo que realice la asignación de los partidos a los horarios de forma que maximice
la audiencia.

• Los horarios disponibles se conocen a priori y son los siguientes:
Viernes 20
Sábado 12,16,18,20
Domingo 12,16,18,20
Lunes 20

En primer lugar se clasifican los equipos en tres categorías según el numero de
seguidores( que tiene relación directa con la audiencia). Hay 3 equipos en la
categoría A, 11 equipos de categoría B y 6 equipos de categoría C.
• Se conoce estadísticamente la audiencia que genera cada partido según los equipos
que se enfrentan y en horario de sábado a las 20h (el mejor en todos los casos)
Categoría A Categoría B Categoría C
Categoría A 2 Millones 1,3 Millones 1 Millones
Categoría B 0.9 Millones 0.75 Millones

Si el horario del partido no se realiza a las 20 horas del sábado se sabe que se reduce
según los coeficientes de la siguiente tabla
• Debemos asignar obligatoriamente siempre un partido el viernes y un partido el lunes
Viernes Sábado Domingo Lunes
12h - 0.55 0.45 -
16h - 0.7 0.75 -
18h - 0.8 0.85 -
20h 0.4 1 1 0.4

Es posible la coincidencia de horarios pero en este
caso la audiencia de cada partido se verá afectada y
se estima que se reduce en porcentaje según la
siguiente tabla dependiendo del número de
coincidencias.
(Las tablas del enunciado no se han podido copiar bien, pero lo demás es todo el enunciado)

....

(*) La respuesta es obligatoria





                                        

(*)**¿Cuantas posibilidades hay sin tener en cuenta las restricciones?**<br>



**¿Cuantas posibilidades hay teniendo en cuenta todas las restricciones.**




Respuesta

En el caso de no tener en cuenta las restricciones, cualquier partido podría emitirse en cualquier horario, es decir, será 10^n.

In [ ]:
def posibilidades_no_restricciones(numero_franjas, numero_partidos):
  return numero_partidos**numero_franjas

print(f'Las posibilidades sin restricciones son {posibilidades_no_restricciones(10, 10)}')

Las posibilidades sin restricciones son 10000000000


En el caso de tener en cuenta las restricciones, en las 10 franjas de horario que se tienen disponibles hay que eliminar los casos, que por las restricciones no son posibles. En este caso son los viernes a las 20h y el lunes a las 20h. Además también existe la opción de que las dos restricciones sucedan a la vez, es decir, que no haya partidos ni viernes a las 20 ni lunes a las 20.

In [ ]:
def posibilidades_si_restricciones(numero_partidos):
  return (10**numero_partidos)+(8**numero_partidos)-(9**numero_partidos)-(9**numero_partidos)

print(f'Las posibilidades con restricciones son {posibilidades_si_restricciones(10)}')

Las posibilidades con restricciones son 4100173022


**Modelo para el espacio de soluciones**<br>
(*) **¿Cual es la estructura de datos que mejor se adapta al problema? Argumentalo.(Es posible que hayas elegido una al principio y veas la necesidad de cambiar, arguentalo)**


Respuesta

La estructura de datos que mejor se adapta a este problema es una representación basada en listas (arrays) y diccionarios simples, donde cada partido se modela mediante su audiencia base y cada solución se representa como un vector de asignaciones a horarios.

En primer lugar, los partidos pueden representarse mediante una lista de números reales que contienen la audiencia base asociada a cada enfrentamiento según las categorías de los equipos. no es necesario almacenar los nombres de los equipos ni sus categorías durante la optimización, ya que el algoritmo únicamente utiliza la audiencia base para calcular la función objetivo. Por tanto, una lista unidimensional de tamaño n, donde n es el número de partidos de la jornada, es suficiente y además óptima en términos de memoria y acceso.

En segundo lugar, los horarios disponibles pueden representarse mediante una lista fija de diez posiciones, cada una asociada a un identificador (por ejemplo, viernes 20h, sábado 12h, etc.), y un diccionario que almacena el coeficiente multiplicador correspondiente a cada horario. Esto se ve de manera muy visual en la celda siguiente:

In [ ]:
#lista fija de horarios (índice = posición)
SLOTS = [
    "V20",  # 0
    "S12",  # 1
    "S16",  # 2
    "S18",  # 3
    "S20",  # 4
    "D12",  # 5
    "D16",  # 6
    "D18",  # 7
    "D20",  # 8
    "L20"   # 9
    ]

#diccionario con coeficientes
COEF = {
    "V20": 0.4,
    "L20": 0.4,
    "S12": 0.55,
    "S16": 0.7,
    "S18": 0.8,
    "S20": 1.0,
    "D12": 0.45,
    "D16": 0.75,
    "D18": 0.85,
    "D20": 1.0
}

En cuanto a la representación de la solución,
esta se representa mediante una lista de enteros de tamaño n, donde cada posición indica el índice del horario asignado al partido correspondiente.

Esta representación es especialmente eficiente porque transforma el problema en una simple asignación de valores enteros dentro del rango fijo. Además, permite calcular fácilmente el número de coincidencias por horario mediante un vector auxiliar de conteo, también de tamaño diez.

In [ ]:
#asignación[i] = índice del slot asignado al partido i

asignacion = [4, 0, 7, 3, 8, 9, 2, 4, 1, 6]
#esto significa que, por ejemplo, el partido 0 se asigna al índice del slot 4, es decir, a S20.

#conteo coincidencias:
conteo = [0]*10
for s in asignacion:
    conteo[s] += 1

**Según el modelo para el espacio de soluciones**<br>
(*)***¿Cual es la función objetivo?***

(*)***¿Es un problema de maximización o minimización?***

Respuesta

La función objetivo del problema consiste en maximizar la audiencia total generada en una jornada, teniendo en cuenta tanto el atractivo intrínseco de cada partido como el efecto del horario asignado y las penalizaciones derivadas de coincidencias con otros encuentros.


Así pues, la función sería:

f(x) = ∑ᵢ bᵢ · c_{h(i)} · p(k_{h(i)})

donde:

bᵢ es la audiencia base del partido i

c_{h(i)} es el coeficiente del horario asignado al partido i

k_{h(i)} es el número de partidos programados en ese mismo horario

p(k) es el factor de penalización asociado a que haya k partidos simultáneos (según la tabla dada en el problema)

Por ende, se trata un problema de maximización ya que la función objetivo es maximizar la audiencia durante la semana. Para ello, hay que ubicar los partidos en distintas zonas horarias y de esta manera se pueda distribuir la audiencia.

**Diseña un algoritmo para resolver el problema por fuerza bruta**

Para resolverlo por fuerza bruta, la idea es recorrer todas las asignaciones posibles, filtrando las que no tengan V20 y L20 (restricción), calculando la audiencia total y guardando la mejor.

In [ ]:
import itertools, random

#datos
SLOTS = ["V20","S12","S16","S18","S20","D12","D16","D18","D20","L20"]
COEF  = {"V20":0.4,"L20":0.4,"S12":0.55,"S16":0.7,"S18":0.8,"S20":1.0,
         "D12":0.45,"D16":0.75,"D18":0.85,"D20":1.0}

#penalización según num de coincidencias (c = k-1). Devuelve multiplicador.
PEnALIZACIOn_COInCIDEnCIAS = {0:1.0, 1:0.75, 2:0.55, 3:0.40, 4:0.30, 5:0.25, 6:0.22, 7:0.20, 8:0.20}
def pen_from_k(k):  # k partidos en el slot
    if k <= 0: return 0.0
    c = k - 1
    return PEnALIZACIOn_COInCIDEnCIAS.get(c, 0.20)

#bases espectadores (en millones) por categorías
BASE_PAIR = {
    ("A","A"):2.0, ("A","B"):1.3, ("B","A"):1.3,
    ("A","C"):1.0, ("C","A"):1.0,
    ("B","B"):0.9, ("B","C"):0.75, ("C","B"):0.75,
    ("C","C"):0.47
}

#selection sort de índices
def selection_order_desc(valores):
    indice = list(range(len(valores)))
    for i in range(len(indice)-1):
        mejor = i
        for j in range(i+1, len(indice)):
            if valores[indice[j]] > valores[indice[mejor]]:
                mejor = j
        indice[i], indice[mejor] = indice[mejor], indice[i]
    return indice

#evaluar una asignación usando la fórmula por slots
def asignacion_puntuacion(assign, bases):
    suma_pesos = [0.0]*10
    contador  = [0]*10
    #generar las asignaciones posibles
    for i, s in enumerate(assign):
        w = bases[i]*COEF[SLOTS[s]]
        suma_pesos[s] += w
        contador[s]  += 1
    #se aplica la penalizacion del solapamiento y se suma todo.
    total = 0.0
    for s in range(10):
        total += pen_from_k(contador[s]) * suma_pesos[s]
    return total

#Fuerza bruta alg.
def brute_force(bases):
    n = len(bases)
    mejor_valor, mejor_a = -1.0, None
    for a in itertools.product(range(10), repeat=n):
        if 0 not in a or 9 not in a:  # es la restriccion, al menos debe de haber un partido en 0 y en 9
            continue
        val = asignacion_puntuacion(a, bases) #se evalua la condicion
        if val > mejor_valor: #si encuentra un valor mejor, se actualiza.
            mejor_valor, mejor_a = val, a
    return mejor_valor, mejor_a


**Calcula la complejidad del algoritmo por fuerza bruta**

Respuesta

La complejidad temporal del algoritmo de fuerza bruta es O(n*10^n), es decir, O(10^n), una complejidad exponencial. Por eso el problema se vuelve intratable cuando aumenta el número de partidos.

(*)**Diseña un algoritmo que mejore la complejidad del algortimo por fuerza bruta. Argumenta porque crees que mejora el algoritmo por fuerza bruta**

Respuesta

Vamos a implementar un algoritmo de ramificación y poda, con el objetivo de mejorar al algoritmo de fuerza bruta.

Mientras que la fuerza bruta recorre sistemáticamente todas las posibles asignaciones, la ramificación y poda explora el mismo espacio de soluciones potenciales, pero lo hace de manera estructurada, construyendo soluciones parciales y evaluando en cada paso si merece la pena seguir profundizando en esa rama. La mejora se produce, por tanto, porque en muchos casos grandes partes del árbol de búsqueda quedan descartadas en niveles superiores, sin necesidad de llegar a las hojas.

In [ ]:
#ramificacion y poda
def branch_and_bound(bases):
    n = len(bases)

    #exploramos primero los partidos "más importantes"
    order = selection_order_desc(bases) #obtener los indices de mayor a menor
    valores_base_ordenados = [bases[i] for i in order] #reordenar las bases

    #cota rápida (optimista): a lo que falte le ponemos coef 1 y sin penalización => +sum(bases_restantes)
    cota_restante = [0.0]*(n+1)
    for i in range(n-1, -1, -1):
        cota_restante[i] = cota_restante[i+1] + valores_base_ordenados[i] #cota optimista, si quedan partidos libres, aportan valor a su base.

    #estado incremental por slots
    contador  = [0]*10
    suma_pesos = [0.0]*10
    asignacion_ordenada = [0]*n
    total_actual = 0.0

    mejor_valor, mejor_asignacion = -1.0, None
    nodos = 0

    def agregar_a_slot(s, w):
        # actualiza total con delta por cambio de pen(k) y suma_pesos
        nonlocal total_actual
        antigua_k, antigua_suma = contador[s], suma_pesos[s]
        antigua_contribucion = pen_from_k(antigua_k) * antigua_suma #antigua contribucion a slot
        nueva_k, nueva_suma = antigua_k + 1, antigua_suma + w #nuevo estado tras quitar el partido
        nueva_contribucion = pen_from_k(nueva_k) * nueva_suma #nueva contribucion a slot
        contador[s], suma_pesos[s] = nueva_k, nueva_suma #actualiza el estado de slot
        total_actual += (nueva_contribucion - antigua_contribucion)

    def quitar_de_slot(s, w):
        nonlocal total_actual
        antigua_k, antigua_suma = contador[s], suma_pesos[s]
        antigua_contribucion = pen_from_k(antigua_k) * antigua_suma #antigua contribucion a slot
        nueva_k, nueva_suma = antigua_k - 1, antigua_suma - w #nuevo estado tras quitar el partido
        nueva_contribucion = pen_from_k(nueva_k) * nueva_suma #nueva contribucion a slot
        contador[s], suma_pesos[s] = nueva_k, nueva_suma #actualiza el estado de slot
        total_actual += (nueva_contribucion - antigua_contribucion)

    def dfs_busqueda(pos, usa_slot_inicial, usa_slot_final):
        nonlocal mejor_valor, mejor_asignacion, nodos
        nodos += 1

        #poda por cota superior optimista
        if total_actual + cota_restante[pos] <= mejor_valor:
            return

        if pos == n:
            if usa_slot_inicial and usa_slot_final and total_actual > mejor_valor:
                #reconstruir en orden original
                a = [0]*n
                for p, indice in enumerate(order):
                    a[indice] = asignacion_ordenada[p]
                mejor_valor, mejor_asignacion = total_actual, tuple(a)
            return

        restantes = n - (pos + 1)
        base = valores_base_ordenados[pos]

        for s in range(10):
            nuevo_slot_0 = usa_slot_inicial or (s == 0)
            nuevo_slot_9 = usa_slot_final or (s == 9)

            #si ya no quedan partidos después, forzar que se cumpla
            if restantes == 0 and (not nuevo_slot_0 or not nuevo_slot_9):
                continue

            w = base * COEF[SLOTS[s]]
            asignacion_ordenada[pos] = s
            agregar_a_slot(s, w)
            dfs_busqueda(pos+1, nuevo_slot_0, nuevo_slot_9)
            quitar_de_slot(s, w)

    dfs_busqueda(0, False, False)
    return mejor_valor, mejor_asignacion, nodos

(*)**Calcula la complejidad del algoritmo**

Respuesta

En cuanto a la complejidad del algoritmo de ramificación y poda, en el peor caso sigue siendo exponencial. Si tenemos n partidos y m horarios, el árbol completo tiene mⁿ hojas, exactamente igual que en fuerza bruta. Por tanto, la complejidad teórica en el peor caso es O(mⁿ).

Sin embargo, la diferencia fundamental es que, en la práctica, muchas ramas del árbol no se exploran gracias a la poda. Si en un nodo parcial la mejor audiencia que podría alcanzarse (según una cota superior optimista) es menor o igual que la mejor solución completa ya encontrada, esa rama se descarta sin generar todas sus extensiones. Por ello, aunque la cota asintótica no mejora en el peor caso, el número real de nodos explorados suele ser (y será como veremos en apartados posteriores),muchísimo menor que mⁿ.

**Según el problema (y tenga sentido), diseña un juego de datos de entrada aleatorios**

Respuesta

In [ ]:
def random_bases(n=10, seed=0):
  random.seed(seed)
  cats = ["A"]*3 + ["B"]*11 + ["C"]*6
  bases = []
  for _ in range(n):
    c1 = random.choice(cats)
    c2 = random.choice(cats)
    bases.append(BASE_PAIR[(c1,c2)])
  return bases

En primer lugar, la función recibe como parámetro el número de partidos n, lo que permite analizar el comportamiento del algoritmo a distintas escalas del problema.

La utilización de un parámetro seed para inicializar el generador de números aleatorios garantiza la reproducibilidad de los resultados. Sin esta semilla, cada ejecución produciría una instancia distinta y la comparación perdería rigor.

El vector cats está diseñado para reflejar una distribución realista de categorías de equipos. En lugar de asignar las categorías con la misma probabilidad, se construye una lista ponderada donde la categoría A aparece 3 veces, la B 11 veces y la C 6 veces. Esto implica que, al seleccionar aleatoriamente mediante random.choice, la probabilidad de que un equipo pertenezca a cada categoría es proporcional a su frecuencia en la lista.




Para cada partido se seleccionan dos categorías de forma independiente y aleatoria, generando así combinaciones diversas (A-A, A-B, B-C, etc.). Posteriormente, la audiencia base del enfrentamiento se obtiene a través del diccionario BASE_PAIR, que traduce cada pareja de categorías en una estimación numérica de audiencia.

El resultado final es una lista bases de tamaño n que contiene exclusivamente las audiencias base de los partidos generados.

**Aplica el algoritmo al juego de datos generado**

Respuesta

(Tarda aproximadamente 2 mins.)

In [ ]:
# n=10
#Probaremos directamente con alg. ramif. y poda dado que por fuerza bruta tarda mucho y hay riesgo de que se quede colgado.
bases10 = random_bases(n=10, seed=42)
mejor10, a10, nodos10 = branch_and_bound(bases10)
print("Para n=10 el mejor de ramificación y poda es:", mejor10, "con los nodos:", nodos10, "y una asignación:", a10)

#Vamos a probar con un n más pequeño para comparar ambos (fuerza bruta y B&B)
# n=6
bases6 = random_bases(n=6, seed=7)
bval, ba = brute_force(bases6)
bbval, bba, nodos = branch_and_bound(bases6)
print("n=6 ruta:", bval, ba)
print("n=6 B&B :", bbval, bba, "nodos:", nodos)

Para n=10 el mejor de ramificación y poda es: 7.9729999999008605 con los nodos: 63028641 y una asignación: (8, 1, 2, 0, 3, 7, 4, 5, 9, 6)
n=6 ruta: 4.53 (0, 4, 7, 3, 8, 9)
n=6 B&B : 4.5299999999997755 (3, 4, 8, 0, 7, 9) nodos: 59294


En primer lugar, para n = 6 se observa que la audiencia obtenida mediante fuerza bruta (4.53) y la obtenida mediante ramificación y poda (4.5299) coinciden. La ligera diferencia numérica se debe únicamente a errores de representación en coma flotante, no a diferencias reales en la audiencia. Esto confirma que el algoritmo de ramificación y poda ha encontrado el óptimo global, ya que coincide con el resultado de fuerza bruta, que evalúa exhaustivamente todas las combinaciones posibles.

Además, las asignaciones encontradas son distintas, pero producen la misma audiencia total. Esto indica que el problema puede tener múltiples soluciones óptimas equivalentes, algo habitual en problemas de optimización combinatoria.

Para n = 10 la situación es aún más ilustrativa. El espacio total teórico sería 10^10 = 10.000.000.000 soluciones posibles. Evidentemente, fuerza bruta sería completamente inuevo_slot_0iable en este caso. El algoritmo de ramificación y poda ha explorado 63.028.641 nodos, encontrando una audiencia óptima de 7.97 millones.

Aunque 63 millones de nodos puede parecer un número elevado, representa únicamente una fracción muy pequeña del total posible (aproximadamente un 0,63 % del espacio completo). Es decir, se ha evitado explorar más del 99 % de las combinaciones.

Estos resultados permiten justificar que, aunque la complejidad teórica en el peor caso sigue siendo exponencial, la complejidad real en la práctica se reduce de manera drástica gracias a la poda.

Por último, el hecho de que para n = 6 el resultado coincida exactamente con fuerza bruta garantiza que el método no es heurístico, sino exacto. Por ello, se mantiene la garantía de optimalidad global mientras se reduce sustancialmente el número de nodos explorados.

En resumen, los resultados confirman que la ramificación y poda mejora de forma muy notable el rendimiento respecto a fuerza bruta, especialmente a medida que crece el tamaño del problema, manteniendo al mismo tiempo la garantía de encontrar el óptimo global.

**Enumera las referencias que has utilizado(si ha sido necesario) para llevar a cabo el trabajo**

Respuesta

**Describe brevemente las lineas de como crees que es posible avanzar en el estudio del problema. Ten en cuenta incluso posibles variaciones del problema y/o variaciones al alza del tamaño**

Respuesta

Una opción posible es añadir alguna nueva restricción como que las finales de las grandes competiciones solo se puedan hacer en horarios donde hay máxima audiencia o establecer la norma que hay que coordinar el horario para que puedan realizar un entrenamiento de una semana mínimo entre cada partido que se juega. De esta manera se podría separar el problema en distintos bloques y poder tratarlo de mejor forma.


Además, se puede crear una tabla con las puntuaciones de los equipos para hacer un ranking entre los equipos que se presenten a la competición. Para que en la final se encuentren solo los mejores, y esto produzca que la audiencia aumente. También pueden aumentar el número de franjas disponibles y quitar las restricciones de los lunes y los viernes.De esta manera hay menos restricciones y permite que existan más posibilidades de obtener la máxima audiencia en cualquier momento.

